src/mito2/src/worker.rs

```diff
@@ -834,22 +839,23 @@ impl<S: LogStore> RegionWorkerLoop<S> {
         }
     }
 
-    /// Handles `set_readonly_gracefully`.
-    async fn set_readonly_gracefully(
+    /// Handles `set_state_gracefully`.
+    async fn set_role_state_gracefully(
         &mut self,
         region_id: RegionId,
-        sender: oneshot::Sender<SetReadonlyResponse>,
+        region_role_state: SettableRegionRoleState,
+        sender: oneshot::Sender<SetRegionRoleStateResponse>,
     ) {
         if let Some(region) = self.regions.get_region(region_id) {
             // We need to do this in background as we need the manifest lock.
             common_runtime::spawn_global(async move {
-                region.set_readonly_gracefully().await;
+                region.set_role_state_gracefully(region_role_state).await;
 
                 let last_entry_id = region.version_control.current().last_entry_id;
-                let _ = sender.send(SetReadonlyResponse::success(Some(last_entry_id)));
+                let _ = sender.send(SetRegionRoleStateResponse::success(Some(last_entry_id)));
```

Comment:
Is it possible that the region handles new write requests before the background task setting the region state? Can we ensure the integrity of the data?

Reply:
Yes, in the upper layer. we will convert region to downgrading first, and call flush memtable, finally convert it to readonly.